include/swift/SILOptimizer/Utils/UpdatingInstructionIterator.h

```diff
@@ -0,0 +1,243 @@
+    state().advance();
+    return *this;
+  }
+
+  bool operator==(const Self &rhs) const { return state() == rhs.state(); }
+  bool operator!=(const Self &rhs) const { return !(*this == rhs); }
+};
+
+/// Track iterators that need updating when intructions are added or
+/// deleted. Iterators can be tracked across multiple levels of the call
+/// stack. This registry object must outlive any iterators that it vends.
+class UpdatingInstructionIteratorRegistry {
+  friend UpdatingInstructionIterator;
+  using State = detail::UpdatingInstructionIteratorState;
+
+  /// Track any iterators that need updating. Seldom expect to have more than
+  /// 2-4 of these.
+  ///
+  /// Advancing each iterator moves to the instruction following most recently
+  /// visited instruction. This holds even if the most recent instruction has
+  /// been deleted and new instructions have been added after its original
+  /// position.
+  SmallVector<State, 4> iteratorStates;
+
+  /// Callbacks used when adding/deleting instructions.
+  InstModCallbacks callbacks;
+
+public:
+  UpdatingInstructionIteratorRegistry(
+      InstModCallbacks chainedCallbacks = InstModCallbacks()) {
+
+    // Copy the two std::functions that we need.
+    auto chainedDelete = chainedCallbacks.deleteInstFunc;
+    auto chainedNew = chainedCallbacks.createdNewInstFunc;
+    callbacks = chainedCallbacks
+      .onDelete([this, chainedDelete](SILInstruction *toDelete) {
+        for (auto &iteratorState : iteratorStates) {
+          iteratorState.notifyDelete(toDelete);
+        }
+        if (chainedDelete) {
+          chainedDelete(toDelete);
+          return;
+        }
+        toDelete->eraseFromParent();
+      })
+      .onCreateNewInst([this, chainedNew](SILInstruction *newlyCreatedInst) {
+        for (auto &iteratorState : iteratorStates) {
+          iteratorState.notifyNew(newlyCreatedInst);
+        }
+        if (chainedNew) {
+          chainedNew(newlyCreatedInst);
+        }
+      });
+  }
+
+  InstModCallbacks &getCallbacks() { return callbacks; }
+
+  UpdatingInstructionIterator makeIterator(SILBasicBlock::iterator i) {
+    return {*this, registerIterator(i)};
+  }
+
+  UpdatingInstructionIterator makeIterator(SILInstruction *inst) {
+    return makeIterator(inst->getIterator());
+  }
+
+  llvm::iterator_range<UpdatingInstructionIterator>
+  iteratorRange(SILBasicBlock *bb) {
+    return {makeIterator(bb->begin()), makeIterator(bb->end())};
+  }
+
+private:
+  /// Register an iterator for update and return its ID.
+  unsigned registerIterator(SILBasicBlock::iterator iterator) {
+    unsigned id = iteratorStates.size();
+    iteratorStates.emplace_back(iterator);
+    return id;
+  }
+
+  State &getIteratorState(unsigned id) { return iteratorStates[id]; }
+
+  void retainIterator(unsigned id) { ++iteratorStates[id].refCount; }
+
+  void releaseIterator(unsigned id) {
+    if (--iteratorStates[id].refCount == 0)
+      iteratorStates.erase(iteratorStates.begin() + id);
```

Comment:
That will erase the wrong state if another state was released before with a lower id.